PurchasesCoreSwift/Networking/ETagManager.swift

```diff
@@ -0,0 +1,184 @@
+        }
+        
+        if (eTagInResponse != nil) {
+            if (shouldUseCachedVersion(responseCode: statusCode)) {
+                guard let storedResponse = getStoredHTTPResponse(for: request) else {
+                    if (retried) {
+                        Logger.warn(
+                                String(
+                                        format: Strings.network.could_not_find_cached_response_in_already_retried,
+                                        resultFromBackend
+                                )
+                        )
+                        return resultFromBackend
+                    } else {
+                        return nil
+                    }
+                }
+                return storedResponse
+            }
+            storeStatusCodeAndResponseIfNoError(
+                    for: request,
+                    statusCode: statusCode,
+                    responseObject: responseObject,
+                    eTag: eTagInResponse!)
+        }
+
+        return resultFromBackend
+    }
+
+    @objc public func clearCaches() {
+        if let bundleID = Bundle.main.bundleIdentifier {
+            self.userDefaults.removePersistentDomain(forName: bundleID + ".revenuecat.etags")
+        }
+    }
+
+    private func shouldUseCachedVersion(responseCode: Int) -> Bool {
+        responseCode == HTTPStatusCodes.notModifiedResponseCode.rawValue
+    }
+
+    private func getStoredETagAndResponse(for request: URLRequest) -> ETagAndResponseWrapper? {
+        if let cacheKey = eTagDefaultCacheKey(for: request) {
+            if let data = userDefaults.object(forKey: cacheKey) as! Data? {
+                do {
+                    return try ETagAndResponseWrapper(with: data)
+                } catch {
+                }
+            }
+        }
+
+        return nil
+    }
+
+    private func getStoredHTTPResponse(for request: URLRequest) -> HTTPResponse? {
+        if let storedETagAndResponse = getStoredETagAndResponse(for: request) {
+            return HTTPResponse(
+                    statusCode: storedETagAndResponse.statusCode,
+                    responseObject: storedETagAndResponse.responseObject)
+        }
+
+        return nil
+    }
+
+    private func storeStatusCodeAndResponseIfNoError(for request: URLRequest,
+                                                     statusCode: Int,
+                                                     responseObject: [String: Any]?,
+                                                     eTag: String) {
+        if statusCode != HTTPStatusCodes.notModifiedResponseCode.rawValue &&
+                   statusCode < HTTPStatusCodes.internalServerError.rawValue &&
+                   responseObject != nil {
+            if let cacheKey = eTagDefaultCacheKey(for: request) {
+                let eTagAndResponse =
+                        ETagAndResponseWrapper(eTag: eTag, statusCode: statusCode, responseObject: responseObject!)
```

Comment:
same story with force unwraps as with force casts, we should avoid them at all costs

Reply:
I was confused about this one. The conditional has a != nil, why the compiler doesn't know it is not supposed to be nil?

Reply:
because swift is dumb 😂  or at least, certainly dumber in this case than Dart, which would be able to figure this out. 

Easy way around it is to use lets

i.e.: 

```swift
if statusCode != HTTPStatusCodes.notModifiedResponseCode.rawValue &&
    statusCode < HTTPStatusCodes.internalServerError.rawValue,
   let responseObject = responseObject,
   let cacheKey = eTagDefaultCacheKey(for: request) {
    let eTagAndResponse = ETagAndResponseWrapper(eTag: eTag,
                                                 statusCode: statusCode,
                                                 responseObject: responseObject)
    if let dataToStore = eTagAndResponse.asData() {
        userDefaults.set(dataToStore, forKey: cacheKey)
    }
}
```

Reply:
but to be fair to swift, if `responseObject` had been a reference type, though, you'd have no guarantee that it isn't nil by line 107, even if you did check for it before, since it could have been modified on a different thread. 
in this case, `responseObject` is a dict, which is passed as a copy, buuuuut you could always do `responseObject = nil` right before this line and break the assumption.

Reply:
🤔 thinking about it more, `responseObject = nil` won't compile, since it gets passed as a constant. So, yeah, it's swift being dumb.